### ISSUE TITLE
[css-text] Provide CSS support for automatic spacing (or other separators) in text input field

### ISSUE THREAD

**Issue by marvindanig** [w3c/csswg-drafts#4875]
This is a feature request to allow styling of the text input field for formatted inputs like credit card or date or phone number with an area code.  Currently this is [not possible](https://twitter.com/AmeliasBrain/status/1238879132475219968) without using javascript. 

Shown below is an example where a credit card number is split into groupings of four digits each:

![image](https://user-images.githubusercontent.com/5659924/76688582-84ac2d00-6604-11ea-9a32-3de87328470e.png)

In general, people tend to use comma separation or spaces between large numbers (like a credit card) to pin for accuracy of input. So in some ways this is an accessibility issue as well.  

At this point all of the solutions out there require using [javascript](https://stackoverflow.com/questions/60679224/is-there-a-way-to-format-the-input-field-for-credit-card-info-without-using-java) to format the inputted string by listening to the `onblur` event. There are even libraries from large providers to style the information fields such as this [one](https://stripe.com/docs/stripe-js) from Stripe.  Linked here is the Twitter [thread](https://twitter.com/marvindanig/status/1238885587928723458) that prompted this proposal.

**Comment by AmeliaBR** [comment]
I agree that this is a common design need. I don't have references right now, but I remember reading somewhere that editing the input on blur (to add hard-coded spaces or other separators) can be an accessibility issue.

My rough strawman proposal (from the Twitter thread) was that spacing could be supported by extending the `letter-spacing` property to support a sequence of spaces, like `0 0 0 1ch` for a space after every fourth character.

But that wouldn't help with things like automatically adding dashes to phone numbers or automatically formatting numbers with commas, so maybe a more dedicated number-format approach would be better.

I still think it's useful to talk about this in CSS (as opposed to something like a `format` attribute on an HTML `<input>`), so that the same approach works for custom inputs or general number formatting (e.g., in a data table).

**Comment by Loirooriol** [comment]
Something like a custom [`text-transform`](https://drafts.csswg.org/css-text/#text-transform-property) that replaces every 4th character with itself and a space or dash? The spec already says

> A future level of CSS may introduce the ability to create custom mapping tables for less common text transforms, such as by an @text-transform rule similar to @counter-style

**Comment by Crissov** [comment]
This be #3132 then. Without checking, I think @frivoalʼs proposal could not do this in its current state.

**Comment by tabatkins** [comment]
Yeah, I agree that this is a pretty common and reasonable thing to want. Scoping the design will be interesting, but I suspect it's doable.

**Comment by plinss** [comment]
Note that not all credit cards use a 4/4/4/4 format. Other common formats are 4/4/4, 4/4/5, 4/4/4/2, 4/6/5, 6/11, 6/12, 6/13, and 4/4/4/3. A proper credit card input implementation will look at the card digits to determine the issuer and switch formats as the user is typing. Similar issue for phone numbers. Whatever we do here has to be dynamic, at the least letting the author change formats in a non-destructive manner in response to input events.

**Comment by AmeliaBR** [comment]
@plinss That's a good point. But considering that all existing implementations use dynamic JS anyway, it might still be an improvement if they could preserve the distinction between the input characters and the formatting adjustments.

**Comment by plinss** [comment]
I’m not saying that odd cases like credit cards and phone numbers should be doable without JS entirely, a small amount of code in an oninput to swap formats is fine, and IMO desirable because there’s no way we can predict all the ways this will be used. Just that an author should be able to swap the formatter dynamically without interfering with user input.

I wrote a CC input that does the dynamic formatting by adjusting the value, and it was a PITA and still likely misbehaves in some circumstances. I’d love to be able to throw that away for something that works gracefully.

**Comment by tabatkins** [comment]
Ah yup, a script that just adjusts `input.style.formatPattern` on the fly would be big improvement over that.

**Comment by ByteEater-pl** [comment]
Per the principle of least power, could regular expressions be a viable alternative to scripting? (Events could be fired anyway in case the author wants to do something more than declaratively possible, as with animations and much other CSS stuff.)

**Comment by fantasai** [comment]
I think this request makes a lot of sense, but I think it probably belongs at the HTML layer. @ByteEater-pl’s suggestion to use regex makes sense: we already have a `pattern` attribute to validate input, maybe a `format` attribute that, using captured substrings, describes the output formatting?

**Comment by ByteEater-pl** [comment]
While validation is indeed a host language concern, displaying transformed text seems more like a generalization of text-transform and applicable regardless of if it's HTML being styled or something else.

**Comment by faceless2** [comment]
This seems to be another variation of "I need to apply a transform to text styled with CSS to modify spacing, or change it in a way more complex than what we can currently do with `text-transform`".

It's a problem that keeps coming up. it's been solved so many times and in so many ways, it seems odd that a more generalised "text-fixup" option isn't part of CSS. Here's my partial list of references.

https://www.w3.org/TR/css-text-4/#text-spacing-property
https://www.w3.org/Style/2013/paged-media-tasks#text-fix
https://www.princexml.com/doc/css-props/#prop-prince-text-replace
https://www.antennahouse.com/product/ahf66/ahf-ext.html#axf.text-replace
https://specs.rivoal.net/css-custom-tt/
https://books.idea.whatwg.org/#character-substitution
#4246

All of these proposals relate primarily* to changing the _formatted_ value of the text, rather than the _logical_ value: this proposal included (I don't think there request is that spaces are included as part of the submitted credit card number, just that they're included when the value is displayed on screen).

For this reason I think it's a styling issue and part of CSS's domain.

(*) this is a slightly blurred distinction - first because although that seems to be how it's normally used, it doesn't have to be (see for example https://lists.w3.org/Archives/Public/www-style/2008Sep/0013.html), and second because of https://github.com/w3c/csswg-drafts/issues/3775

**Comment by AmeliaBR** [comment]
I am currently filling out a PDF form & I noticed that after I edit, it is automatically formatting the numbers with commas for thousands separators. Here are Adobe docs for the types of formats that are supported:
https://helpx.adobe.com/ca/acrobat/using/pdf-form-field-properties.html#format_tab_for_form_field_properties

Might be worth further investigation to see if that model could be adapted in a consistent yet CSS-compatible way.

**Comment by faceless2** [comment]
In PDF that's all done with JavaScript.

PDF has "keypress" events which fire as you type into the field, and a "format" event which fires when the value of the field changes. It also makes a distinction between the actual value of the field and the _formatted_ value as it appears on screen - the two can be different, which is how you can get commas in the numbers but not in the submitted value.

**Comment by TimeTerminal** [comment]
It's been a year and half since this issue was opened. Any further discussion or proposal on this?

**Comment by SebastianZ** [comment]
So far there wasn't any further discussion, obviously.

I really like the idea of an `@text-transform` rule as proposed in #3132 and mentioned earlier by @Crissov. Though that proposal would need to be extended to also cover adding characters at specific positions and other more complex transformations than simple character replacements.

Sebastian

**Comment by Crissov** [comment]
`text-transform`, for now, is about using the glyphs of other characters, so it could only apply if there already was a grouping character. Therefore, I think this feature should be handled by a different property.

Transformation between number formats like these would be in scope for `text-transform`:

* `123.456,789`
* `123,456.789`
* `123’456·789`
* `123 456.789`

… as would be transformation between code formats like these:

* `123-456-789`
* `123_456_789`
* `123:456:789`
* `123/456/789`
* `123 456 789`
* `123⁣456⁣789` U+2063
* `123​456​789` U+200B
* `123	456	789` U+0009

… and the digit style itself of course #9877 ;)

**Comment by tszyuloveyou** [comment]
May I know which time can use.

**Comment by tomwhoiscontrary** [comment]
I wonder if the interaction with copy and paste should to be considered.

For example, in the UK, bank accounts have sort codes which are conventionally written 01-23-45. I'd like to be able to type into a sort code field using only digits, and have it displayed conventionally. But i would also like to be able to copy from a conventionally formatted source, say a PDF invoice, and paste that into a sort code field, hyphens and all, and get the right result.

Then what happens if i select all and copy from such a field? Should i get back what i typed, or what i see? What if the application populated the field, and i never typed anything?

**Comment by Crissov** [comment]
There now is a [[css-forms] draft proposal](https://drafts.csswg.org/css-forms-1/) where this might also belong. Alas, it currently only has pseudo-elements for separatists in date and time into fields.

**Comment by nt1m** [comment]
I would love this done at the HTML level. Maybe a `format` HTML attribute as suggested above would be reasonable.

**Comment by SebastianZ** [comment]
One advantage of a CSS solution is that it could also apply to elements other than text input fields. This allows to

**Comment by SebastianZ** [comment]
Sorry, hit the wrong button and too early. Adjusted my previous comment.

Sebastian
